### ISSUE TITLE
fix perf regression from parsing documentation comments

### ISSUE THREAD

**Issue by sigmundch** [dart-lang/sdk#30899]
We saw a regression in fasta when we added parsing for documentation comments. We should figure out where the regression came from (could be from the extra parsing itself, or from the way we store the documentation comments in the IR) and ensure that if we don't need the comments, we don't pay a cost for them.

**Comment by scheglov** [comment]
I looked on Golem `Benchmark?benchmark=scan_dart2js&team=dart-frontend` and don't see the regression. Is this the correct benchmark?

**Comment by scheglov** [comment]
We parse comments only when `KernelTarget` is told so via its constructor parameter `includeComments`.

As I can see, `KernelDriver` uses `true`, and `generateKernelInternal` from `kernel_generator_impl.dart` uses `false`. Fasta tool also uses `false` in `CompileTask::createKernelTarget`. So, there should be no cost from documentation comments in benchmarks, other than from the field and attempt to serialize `null` strings.

**Comment by sigmundch** [comment]
I saw this under the `kernel_e2e_dart2js` benchmark.

I'm trying to search back when documentation parsing was added, and I might be wrong that it wasn't when it was first introduced, but when we started handling "///" and "/*" in https://github.com/dart-lang/sdk/commit/2d01f1f8d9c992fb3da442b6550213d2555f2f3c.

This commit corresponds to revision 61638 on golem. It doesn't show up in the NanosecondsPerChar metric, but does on the RunTimeRaw.

**Comment by sigmundch** [comment]
mmm. it seems that the change also affected the scan_dart2js benchmark, but I don't quite understand why.

**Comment by scheglov** [comment]
Yes, it is suspicious that `scan_dart2js` RunTimeRaw also jumped `14%`, but on a different https://github.com/dart-lang/sdk/commit/09ff1b56f79ba855a5ef7149be2f7bc6536154ec, i.e. 61636. Both i32 and x64. And `parse_dart2js` also jumped not on my 61638, but on previous 61636.

Actually even `kernel_gen_e2e_dart2js` shows that 61636 causes `12-15%` slowdown on `dart2js-compile - linux-ia32 - Intel Core i5` and `dart2js-compile - linux-x64 - Intel Core i5`. True, `d2fe-fasta - linux-ia32 - Intel Core i5` shows that 61638 causes the slowdown, but there is no 61636 on this graph - there is a jump from 61631 to 61638, we don't see performance of intermediate changes.

**Comment by sigmundch** [comment]
Ah! Thanks for looking into it - that makes complete sense now.

The earlier change adds a lot of code to dart2js (this is where we added code to invoke the FE from dart2js), so the program we are benchmarking got bigger. That explains why both benchmarks got a 10% runtime increase, and it also explains why there was no change in the nanosecondsPerChar metric (the speed didn't change, just the size of the input).

The benchmark bots have heuristics to only run a benchmark when it knows it is affected by a change. My guess is that it incorrectly assumed that a change to dart2js was irrelevant here (technically it didn't affect the FE, but it affect the data we gave to the benhmark). I'm guessing that your CL was the first FE CL that got triggered after that.

I'll follow up with the folks working on the performance bots.

As far as this bug is concerned I'm not sure if there is much more for us to do, so I'll close it.
